### PR TITLE
Autocreate product / engagement during (re)import

### PR DESCRIPTION
**Pull request by valentijnscholten**
This PR finalizes the `smartness` to the importer logic to allow the API wrappers and upload scripts to be more simple.
Clients using the API now only need to provide a product / engagement name and can use the `reimport-scan` endpoint, even for first time uploads. This should bring the 'easy of use' of API imports on par with for example Dependency Track or SonarQube where you can (re)upload a report by simply specifying some product name / project name.

The `import-scan` APi endpoint now has a new parameters `auto_create_context` which, if set to `True`, will instruct the import logic to automatically create an `Engagement` (and `Product` and `Product_Type`) if needed. The existing parameters with `engagement_name` and `product_name` from my previous "Smart Import" PR are used. An extra parameter `product_type_name` is added to allow creation of Product Types as well.

The `reimport-scan` now also supports autocreation of these objects, but it also supports first time scan uploads. The idea is that a (CI/CD) API client script can now just call the `reimport-scan` endpoint and the endpoint will figure out if it is a reupload as usual or a first time upload. For a first time upload a new Test is created using the same logic as the `import-scan` upload.

For all creations the proper permission checks are done. Currently the `API_Importer` role doesn't have enough permissions to do all this, so people should use the global `Maintainer` or `Owner` roles. We could extend the `API_Importer` role with more permissions, or create an `API_Importer++` role. But it will soon start to look like a global `Maintainter` role I think, but I'm open to suggestions or future PRs.

While I was writing the `docs` for this PR, I did a little revamp by putting the documentation about Import/Reimport on its own page leaving the existing page dedicated to list the supported parsers / reports.

I guess this PR is not a "quick review", so I made sure there are lots of unit tests to cover the new feature(s).

### CODE REVIEW

File: dojo/api_v2/permissions.py [hunk 1]
```diff
@@ -181,24 +183,26 @@ def has_permission(self, request, view):
...
         # permission check takes place before validation, so we don't have access to serializer.validated_data()
         # and we have to validate ourselves unfortunately
 
-        _, _, _, _, _, product_name = get_import_meta_data_from_dict(request.data)
+        _, _, _, _, _, product_name, _, _ = get_import_meta_data_from_dict(request.data)
         product = get_target_product_if_exists(product_name)
         if not product:
             product_id = get_product_id_from_dict(request.data)
-            product = get_target_product_by_id_if_exists(product_id)
+            product = get_target_product_by_id_if_exsits(product_id)
```
Comment (Maffooch) [inline dojo/api_v2/permissions.py:206]: Stefan fixed my oopses here. Please pull in current dev changes

File: dojo/api_v2/serializers.py [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 from drf_spectacular.utils import extend_schema_field
 from drf_yasg.utils import swagger_serializer_method
+from rest_framework.exceptions import NotFound
 from rest_framework.fields import DictField, MultipleChoiceField
 
 from dojo.endpoint.utils import endpoint_filter
-from dojo.importers.reimporter.utils import get_target_engagement_if_exists, \
-    get_target_product_if_exists, get_target_test_if_exists, get_target_product_by_id_if_exists
+from dojo.importers.reimporter.utils import get_or_create_engagement, get_target_engagement_if_exists, get_target_product_by_id_if_exsits, \
+    get_target_product_if_exists, get_target_test_if_exists
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:9]: Same issue

File: dojo/api_v2/serializers.py [hunk 3]
```diff
@@ -1480,11 +1520,11 @@ def save(self):
         create_tags = data['create_tags']
         create_dojo_meta = data['create_dojo_meta']
 
-        _, _, _, _, _, product_name = get_import_meta_data_from_dict(data)
+        _, _, _, _, _, product_name, _, _ = get_import_meta_data_from_dict(data)
         product = get_target_product_if_exists(product_name)
         if not product:
             product_id = get_product_id_from_dict(data)
-            product = get_target_product_by_id_if_exists(product_id)
+            product = get_target_product_by_id_if_exsits(product_id)
```
Comment (Maffooch) [inline dojo/api_v2/serializers.py:1527]: Same issue

File: dojo/importers/reimporter/utils.py [hunk 4]
```diff
@@ -82,14 +84,28 @@ def mitigate_endpoint_status(endpoint_status, user):
     endpoint_status.save()
 
 
-def get_target_product_if_exists(product_name=None):
-    if product_name:
-        return get_object_or_none(Product, name=product_name)
+def get_target_product_type_if_exists(product_type_name=None):
+    if product_type_name:
```
Comment (Maffooch) [inline dojo/importers/reimporter/utils.py:88]: Last issue 🙄

File: dojo/importers/reimporter/utils.py [hunk 5]
```diff
@@ -82,14 +84,28 @@ def mitigate_endpoint_status(endpoint_status, user):
...
+            if product_type_name:
+                if product.product_type.name == product_type_name:
+                    return product
+            else:
+                return product
+
+    return None
+
+
+def get_target_product_by_id_if_exsits(product_id=None):
```
Comment (Maffooch) [inline dojo/importers/reimporter/utils.py:108]: Wait here's another

**Review by Maffooch — COMMENTED** [review]
Love this so far. I have written this same process in automation pipelines through the API with many endpoints. Glad to see it all bundled into the [re]import functions.

**Comment by valentijnscholten** [comment]
Thanks, some merge artifacts as my PR inserts code at places where your previous PR inserted your code. Should be good now (tests running).

**Review by mtesauro — APPROVED** [review]
Approved

**Review by Maffooch — APPROVED** [review]

**Comment by Maffooch** [comment]
Merge when you're ready @valentijnscholten

**Pull request merged**